library/std/src/sys/pal/uefi/helpers.rs

```diff
@@ -142,7 +147,52 @@ pub(crate) unsafe fn close_event(evt: NonNull<crate::ffi::c_void>) -> io::Result
 
 /// Get the Protocol for current system handle.
 /// Note: Some protocols need to be manually freed. It is the callers responsibility to do so.
-pub(crate) fn image_handle_protocol<T>(protocol_guid: Guid) -> Option<NonNull<T>> {
-    let system_handle = uefi::env::try_image_handle()?;
-    open_protocol(system_handle, protocol_guid).ok()
+pub(crate) fn image_handle_protocol<T>(protocol_guid: Guid) -> io::Result<NonNull<T>> {
+    let system_handle = uefi::env::try_image_handle().ok_or(io::const_io_error!(
+        io::ErrorKind::NotFound,
+        "Protocol not found in Image handle"
+    ))?;
+    open_protocol(system_handle, protocol_guid)
+}
+
+pub(crate) fn device_path_to_text(path: *mut device_path::Protocol) -> io::Result<OsString> {
+    let device_path_to_text_handles = locate_handles(device_path_to_text::PROTOCOL_GUID)?;
+    for handle in device_path_to_text_handles {
+        let Ok(protocol) = open_protocol::<device_path_to_text::Protocol>(
+            handle,
+            device_path_to_text::PROTOCOL_GUID,
+        ) else {
+            continue;
+        };
+
+        let path_ptr: *mut r_efi::efi::Char16 = unsafe {
+            ((*protocol.as_ptr()).convert_device_path_to_text)(
+                path,
+                // DisplayOnly
+                r_efi::efi::Boolean::FALSE,
+                // AllowShortcuts
+                r_efi::efi::Boolean::FALSE,
+            )
+        };
+
+        // SAFETY: Error only occurs if path is NULL. Thus no need to dealloc path_ptr
```

Comment:
I think this explanation of why we don't need to dealloc the path on error is useful, but it's not discharging the safety obligiation of `WStrUnits::new`. From that function's doc:

```
/// SAFETY: `lpwstr` must point to a null-terminated wide string that lives
/// at least as long as the lifetime of this struct.
```

Reply:
@nicholasbishop I'm lost do you mean I should add a safety comment about the lifetime of lpwstr?

Reply:
Yes, in part. In more detail:
* I think you should keep the existing comment, but remove the `SAFETY:` part, because leaking memory is not a safety violation.
* Add a new safety comment that "discharges" the requirements of the safety doc of `WStrUnits::new`. (I'm using "discharges" as done in [rfc#2585](https://rust-lang.github.io/rfcs/2585-unsafe-block-in-unsafe-fn.html), not sure if there's a better source for that.) See an example in [std/src/sys/pal/windows/args.rs](https://github.com/rust-lang/rust/blob/35dfc67d94c47a6c6ae28c46e7dc1c547f772485/library/std/src/sys/pal/windows/args.rs#L36).

So putting it all together, something like this:
```rust
// SAFETY: `convert_device_path_to_text` returns a pointer to a
// null-terminated UTF-16 string, and that string cannot be deallocated
// prior to dropping the `WStrUnits`, so it's safe for `WStrUnits` to
// use.
let path_len = unsafe {
    WStrUnits::new(path_ptr)
        // The error path can only be taken if `path_ptr` is null, so no
        // need to deallocate `path_ptr` in that case.
        .ok_or(io::const_io_error!(io::ErrorKind::InvalidData, "Invalid path"))?
        .count()
};
```